### PR TITLE
AKU-559: Further async issue resolving

### PR DESCRIPTION
**Pull request by draperd**
This PR (hopefully the last!) relates to https://issues.alfresco.com/jira/browse/AKU-559... or rather the fallout that the most recent changes created.  I observed that the previous fix had introduced further issues where a risk of asynchronous loading of modules might occur. This typically occurred when form controls were added within form layouts such that attempts would be made to update all of the form controls before they were ready. I've re-worked the CoreWidgetProcessing mixin to ensure that we can reliably make multiple calls to processWidgets and have added a new unit test that reproduced the failure and ensured that it is fixed. All the other updates relate to making existing widgets work in this amended style.

### CODE REVIEW

File: aikau/src/main/resources/alfresco/core/CoreWidgetProcessing.js [hunk 1]
```diff
@@ -51,6 +51,110 @@ define(["dojo/_base/declare",
...
+       *
+       * @instance
+       * @type {string}
+       * @default
+       * @since 1.0.36
+       */
+      _countDownLocationPrefix: "_processingCountDown",
+
+      /**
+       * This string is used to identify locations where either processed widgets can be referenced. This location
```
Comment (MartinDoyleUK) [inline aikau/src/main/resources/alfresco/core/CoreWidgetProcessing.js]: ~~either~~

File: aikau/src/main/resources/alfresco/core/CoreWidgetProcessing.js [hunk 2]
```diff
@@ -51,6 +51,110 @@ define(["dojo/_base/declare",
...
+       * as this will only be called when widget processing has completed and is passed an argument of the
+       * widgets that were created.
+       * 
+       * @instance
+       * @param  {string} processWidgetsId The ID mapped to the original processWidgets call
+       * @return {object[]|promise} The array of processed widgets or a promise of them
+       * @since 1.0.36
+       */
+      getProcessedWidgets: function alfresco_core_CoreWidgetProcessing__getProcessedWidgets(processWidgetsId) {
+         var r;
```
Comment (MartinDoyleUK) [inline aikau/src/main/resources/alfresco/core/CoreWidgetProcessing.js]: Why `r`? This doesn't seem a great variable name ...

File: aikau/src/main/resources/alfresco/core/CoreWidgetProcessing.js [hunk 3]
```diff
@@ -113,30 +220,12 @@ define(["dojo/_base/declare",
          else
          {
             this.alfLog("warn", "Could not process widget because it was missing configuration - check 'widgets' array for empty elements", this, index);
-            this._processedWidgetCountdown--;
+            var countdown = lang.getObject(this.getWidgetProcessingLocation(processWidgetsId, this._countDownLocationPrefix), false, this);
+            lang.setObject(this.getWidgetProcessingLocation(processWidgetsId, this._countDownLocationPrefix), countdown - 1, this);
```
Comment (MartinDoyleUK) [inline aikau/src/main/resources/alfresco/core/CoreWidgetProcessing.js]: Duplicated function call to retrieve same string?

```
var location = this.getWidgetProcessingLocation(processWidgetsId, this._countDownLocationPrefix),
   countdown = lang.getObject(location, false, this);
lang.setObject(location, countdown - 1, this);
```

File: aikau/src/main/resources/alfresco/core/CoreWidgetProcessing.js [hunk 4]
```diff
@@ -147,34 +236,23 @@ define(["dojo/_base/declare",
        * @param {string} processWidgetsId An optional ID that might have been provided to map the results of multiple calls to [processWidgets]{@link module:alfresco/core/Core#processWidgets}
        */
       _registerProcessedWidget: function alfresco_core_CoreWidgetProcessing___registerProcessedWidget(widget, index, processWidgetsId) {
-         this._processedWidgetCountdown--;
-         this.alfLog("log", "Widgets expected: ", this._processedWidgetCountdown, this.id);
-         var processedWidgets;
+         // Decrement the count as another widget is registered...
+         var countdown = lang.getObject(this.getWidgetProcessingLocation(processWidgetsId, this._countDownLocationPrefix), false, this) - 1;
+         lang.setObject(this.getWidgetProcessingLocation(processWidgetsId, this._countDownLocationPrefix), countdown, this);
```
Comment (MartinDoyleUK) [inline aikau/src/main/resources/alfresco/core/CoreWidgetProcessing.js]: As above ... "Duplicated function call to retrieve same string?"

File: aikau/src/main/resources/alfresco/core/CoreWidgetProcessing.js (hunk 2 above)
Comment (draperd) [inline aikau/src/main/resources/alfresco/core/CoreWidgetProcessing.js]: r for return, can't use return - I'll call it whatever you want, it makes very little difference to me :)

File: aikau/src/main/resources/alfresco/header/CurrentUserStatus.js [hunk 5]
```diff
@@ -205,9 +206,8 @@ define(["dojo/_base/declare",
        * 
        * @instance
        */
-      setStatusRelativeTime: function alfresco_header_CurrentUserStatus__setStatusRelativeTime()
-      {
-         var relativeTime = (this.userStatusTime == "" || this.userStatusTime == null) ? this.message("status.never-updated") : this.getRelativeTime(this.userStatusTime);
+      setStatusRelativeTime: function alfresco_header_CurrentUserStatus__setStatusRelativeTime() {
+         var relativeTime = (this.userStatusTime === "" || !this.userStatusTime) ? this.message("status.never-updated") : this.getRelativeTime(this.userStatusTime);
```
Comment (MartinDoyleUK) [inline aikau/src/main/resources/alfresco/header/CurrentUserStatus.js]:
```javascript
var relativeTime = (this.userStatusTime === "" || !this.userStatusTime) ? this.message("status.never-updated") : this.getRelativeTime(this.userStatusTime);
```

might be more readable as

```javascript
var relativeTime = this.userStatusTime ? this.getRelativeTime(this.userStatusTime) : this.message("status.never-updated");
```

(if my brain's working right, and that follows the same logic)

File: aikau/src/main/resources/alfresco/layout/css/LeftAndRight.css [hunk 6]
```diff
@@ -1,13 +1,17 @@
 .alfresco-layout-LeftAndRight {
    width: 100%;
-}
 
-.alfresco-layout-LeftAndRight > div > div.left-widgets,
-.alfresco-layout-LeftAndRight > * > div > div.left-widgets {
-   float: left;
-}
+   .alfresco-layout-LeftAndRight__left {
```
Comment (MartinDoyleUK) [inline aikau/src/main/resources/alfresco/layout/css/LeftAndRight.css]: `&__left` ?

File: aikau/src/main/resources/alfresco/layout/css/LeftAndRight.css [hunk 7]
```diff
@@ -1,13 +1,17 @@
...
+      float: left;
+      > * {
+         float: left;
+      }
+   }
 
-.alfresco-layout-LeftAndRight > div > div.right-widgets,
-.alfresco-layout-LeftAndRight > * > div > div.right-widgets {
-   float: right;
+   .alfresco-layout-LeftAndRight__right {
```
Comment (MartinDoyleUK) [inline aikau/src/main/resources/alfresco/layout/css/LeftAndRight.css]: `&__right` ?

File: aikau/src/main/resources/alfresco/core/CoreWidgetProcessing.js (hunk 2 above)
Comment (MartinDoyleUK) [inline aikau/src/main/resources/alfresco/core/CoreWidgetProcessing.js]: widgets? processedWidgets? returnWidgets? single-letters are just really bad for variable names :)

**Comment by draperd** [comment]
Updates pushed

**Pull request merged**